File: 0228. Summary Ranges/0228.java
class Solution {
    public List<String> summaryRanges(int[] nums) {
        // each number is unique
        List<String> ranges = new ArrayList<>();
        List<Integer> consecutive = new ArrayList<>();

        // return empty list if the array is empty
        if (nums == null || nums.length == 0) {
            return ranges;
        }

        for(int i=1; i<nums.length; i++){
            if(nums[i-1]+1 == nums[i]){
                consecutive.add(nums[i-1]);
            }
            else{
                if(consecutive.isEmpty()){
                    // we convert integer to string.
                    ranges.add(String.valueOf(nums[i - 1]));
                }
                else{
                    // we concatenate the first element 
                    // and the last element of the consecutive list with "->".
                    ranges.add(
                        String.valueOf(consecutive.get(0)) 
                        + "->" + 
                        String.valueOf(nums[i - 1])
                    );
                    consecutive.clear();
                }
            }
        }

        // adding remaining elements at the end of the loop
        if(!consecutive.isEmpty()){
            ranges.add(
                String.valueOf(consecutive.get(0)) 
                + "->" + 
                String.valueOf(nums[nums.length - 1])
            );
        } 
        else{
            ranges.add(String.valueOf(nums[nums.length - 1]));
        }
        return ranges;
    }
}

File: 0228. Summary Ranges/0228.py
# Time Complexity: O(n) - Checks each element in the list once.
# Space Complexity: O(1)
class Solution(object):
    def summaryRanges(self, nums):
        """
        :type nums: List[int]
        :rtype: List[str]
        """
        if not nums:
            return []

        ranges = []
        start = nums[0]
        end = nums[0]

        # loop for each element in the list
        for num in nums[1:]:
            if end+1 == num: # if it's a consecutive number
                end = num
            else:
                if start == end: # if it's not a consecutive number
                    ranges.append(str(start))
                else: # append the consecutive numbers to the list
                    ranges.append(str(start) + "->" + str(end))
                start=num
                end=num

        # for the remaining part
        if not start == end:
            ranges.append(str(start) + "->" + str(end))
        else:
            ranges.append(str(start))

 
        return ranges